test_runner/regress/test_layers_from_future.py

```diff
@@ -0,0 +1,216 @@
+    future_layers = get_future_layers()
+    assert len(future_layers) == 1
+    future_layer = future_layers[0]
+    assert isinstance(future_layer, ImageLayerFileName)
+    assert future_layer.lsn == last_record_lsn
+    log.info(
+        f"got layer from the future: lsn={future_layer.lsn} disk_consistent_lsn={ip.disk_consistent_lsn} last_record_lsn={last_record_lsn}"
+    )
+    assert isinstance(env.pageserver_remote_storage, LocalFsStorage)
+    future_layer_path = env.pageserver_remote_storage.layer_path(
+        tenant_id, timeline_id, future_layer
+    )
+    log.info(f"future layer path: {future_layer_path}")
+    pre_stat = future_layer_path.stat()
+
+    # force removal of layers from the future
+    tenant_conf = ps_http.tenant_config(tenant_id)
+    ps_http.tenant_detach(tenant_id)
+    ps_http.configure_failpoints(("before-delete-layer-pausable", "pause"))
+    ps_http.tenant_attach(tenant_id, tenant_conf.tenant_specific_overrides)
+    wait_until_tenant_active(ps_http, tenant_id)
+
+    # Ensure the IndexPart upload that unlinks the layer file finishes, i.e., doesn't clog the queue.
+    def future_layer_is_gone_from_index_part():
+        future_layers = set(get_future_layers())
+        assert future_layer not in future_layers
+
+    wait_until(10, 0.5, future_layer_is_gone_from_index_part)
+
+    # NB: the layer file is unlinked index part now, but, because we made the delete
+    # operation stuck, the layer file itself is still in the remote_storage
+    assert future_layer_path.exists()
+
+    # wait for re-ingestion of the WAL from safekeepers into the in-memory layer
+    # (this happens in parallel to the above)
+    wait_for_last_record_lsn(ps_http, tenant_id, timeline_id, last_record_lsn)
+    # re-do image layer generation
+    # This will produce the same image layer and queue an upload.
+    # However, we still have the deletion for the layer queued, stuck on the failpoint.
+    # So, we have a PUT and a DELETE for the same layer file name racing with each other.
+    # Due to the stuckness of the deletion, the PUT will overtake the DELETE.
+    # Futher, the we have the index part update.
+    # However, these only get scheduled once there are no other inprogress tasks.
+    # But, the stuck layer deletion task is an inprogress task.
+    # So, wait for the file to be replaced
+    ps_http.timeline_compact(tenant_id, timeline_id, force_repartition=True)
+
+    unpause_after_secs = 4
+    start = time.monotonic()
+    while True:
+        post_stat = future_layer_path.stat()
+        if pre_stat.st_ino != post_stat.st_ino:
+            # changed inode <=> overwritten
+            log.warn("observed PUT overtake the stucked DELETE => bug isn't fixed yet")
+            break
+        if time.monotonic() - start > unpause_after_secs:
+            log.info(
+                "a correct implementation would never let the later PUT overtake the earlier DELETE"
+            )
+            break
```

Comment:
Elsewhere in tests we wait for the pausable log line, which is not great given the unconditional logging, but at least it would be less time-dependent.